haskell/private/cc_wrapper.py.tpl

```diff
@@ -0,0 +1,812 @@
+        elif arg.startswith("@"):
+            with open(arg[1:], "r") as rsp:
+                for line in rsp:
+                    for rsp_arg in parse_response_line(line):
+                        yield rsp_arg
+        else:
+            yield arg
+
+
+def parse_response_line(s):
+    # GHC writes response files with quoted lines.
+    return shlex.split(s)
+
+
+def shorten_path(input_path):
+    """Shorten the given path if possible.
+
+    Applies the following transformations if they shorten the path length:
+      - Make path relative to CWD.
+      - Remove redundant up-level references.
+      - Resolve symbolic links.
+
+    Args:
+      input_path: The path to shorten.
+
+    Returns:
+      The shortened path.
+
+    """
+    exists = os.path.exists(input_path)
+    shortened = input_path
+
+    # Try relativizing to current working directory.
+    rel = os.path.relpath(shortened)
+    if len(rel) < len(shortened):
+        shortened = rel
+
+    # Try normalizing the path if possible.
+    norm = os.path.normpath(shortened)
+    if len(norm) < len(shortened):
+        # Ensure that the path is still correct. Reducing up-level references
+        # may change the meaning of the path in the presence of symbolic links.
+        try:
+            if not exists or os.path.samefile(norm, shortened):
+                shortened = norm
+        except IOError:
+            # stat may fail if the path became invalid or does not exist.
+            pass
+
+    # Try resolving symlinks.
+    try:
+        real = os.path.relpath(os.path.realpath(shortened))
+        if len(real) < len(shortened):
+            shortened = real
+    except IOError:
+        # realpath may fail if the path does not exist.
+        pass
+
+    return shortened
+
+
+def rpath_args(rpaths):
+    """Generate arguments for RUNPATHs."""
+    for rpath in rpaths:
+        yield "-Xlinker"
+        yield "-rpath"
+        yield "-Xlinker"
+        yield rpath
+
+
+# --------------------------------------------------------------------
+# Link binary or library
+
+
+def link(output, libraries, rpaths, args):
+    """Execute the link action.
+
+    Args:
+      output: The output binary or library.
+      libraries: Library dependencies.
+      rpaths: The provided rpaths.
+      args: The command-line arguments.
+
+    """
+    if is_darwin():
+        # Reserve space in load commands for later replacement.
+        args.append("-headerpad_max_install_names")
+        rpaths, darwin_rewrites = darwin_shorten_rpaths(
+                rpaths, libraries, output)
+    else:
+        rpaths = shorten_rpaths(rpaths, libraries, output)
+
+    args.extend(rpath_args(rpaths))
+    run_cc(args, exit_on_error=True)
+
+    if is_darwin():
+        darwin_rewrite_load_commands(darwin_rewrites, output)
+
+
+def shorten_rpaths(rpaths, libraries, output):
+    """Avoid redundant rpaths.
+
+    Filters out rpaths that are not required to load any library dependency.
```

Comment:
No, I don't think so. This wrapper is trying to adapt between GHC and Bazel in an unobtrusive way. Redundant rpaths are not an error, but something that can be optimized away, and in the case of macOS something that we have to remove to avoid issues with the MACH-O header size limit.

E.g. GHC will generate lots of rpaths of the form `bazel-out/.../packageXYZ/`. These are redundant in normal linking mode, because rules_haskell passes "better" `$ORIGIN/.../_solib_*/...` rpaths. However, sometimes (e.g. template Haskell) GHC links temporary outputs and doesn't forward rules_haskell's rpaths. In that case we need to use the `bazel-out/...` type rpaths. Also, GHC might pass absolute rpaths, e.g. `/nix/store/...`. These may be redundant if the same library was also imported as a `cc_library` and a corresponding `$ORIGIN/.../_solib_*/...` rpath exists. That's not an error though.

Reply:
So many edge cases :dizzy_face:. But I see how we can’t be strict in that case.